<html>
<head>
<title>DATA 301 - Lab 6</title>
<link href="../assign.css" type=text/css rel=stylesheet>
</head>

<body>

<h1>DATA 301 - Introduction to Data Analytics
<br/>Lab 6: Python Data Analysis</h1>

<hr>

<p>In this lab, we will use Python for automating data analysis tasks. <strong>This lab can be done in pairs. Computer Science majors cannot pair with each other.  You can do it solo if you wish. Pairs are not required.  If working in a pair, only submit from one person with both names on the files.</strong></p>

<h3>Objectives</h3>

<ol>
<li>Use Python to read from a file containing data.</li>
<li>Use Biopython library to manipulate bioinformatics data and request data from public NCBI databases.</li>
<li>Perform data analysis combining data from local and web sources.</li>
<li>Use Python libraries for charting, histograms, k-means clustering, and regression.</li>
</ol>

<h3>Question #1 (15 marks)</h3>

<p>By accident, a bioinformatics lab mixed up DNA sequences in a single data file for two organisms: fruit fly (Drosophila melanogaster) and a grape (Vitis vinifera). Your goal is to figure out which sequences belong to each organism while also learning about the NCBI databases, specifically <a href="http://blast.ncbi.nlm.nih.gov/">BLAST</a>. Create a Python program that reads the sequences from a text file (one per line), performs a BLAST search if not previously done, stores the results of each search in a separate file, and then performs analytics on the search results to help solve our sequence mix-up problem. Details:

<ul>
<li>Put a comment at the top of the Python file called <code>lab6q1</code> with your name and student number. (0.5 mark)</li>
<li>Download the data file called <a href="input.txt">input.txt</a> and place it in your local directory for reading. Each line of the data file contains a DNA sequence. There are 10 sequences in the file.</li>
<li>Read each line of the data file and print to the console the first 20 characters in the sequence and the length of the query sequence. (1 mark)</li>
<li>Use a counter to keep track of how many sequences have been processed.  We will call this <code>count</code>. (0.5 mark)</li>
<li>Using a <code>try-except</code> clause, look for a previously created BLAST result file that should be named <code>dna_lab6_<em>count</em>.xml</code>. For example, <code>dna_lab6_1.xml</code>. Start the number from 1. (2 marks)</li>
<li>If the file exists, open it. Print <code>Using saved file.</code> (1 mark)</li>
<li>If the file does not exist or there is an error, perform a BLAST search using the sequence given.  Print <code>Performing online BLAST search.</code> Note that you will have to import some modules from Biopython which are not already installed with Anaconda.  To install with Anaconda, run this command: <pre>conda install -c https://conda.anaconda.org/anaconda biopython</pre>  Otherwise, <a href="http://biopython.org/wiki/Download">see these download/install instructions</a>.</li>
<li>Libraries to import:
<pre>
from Bio.Blast import NCBIWWW
from Bio.Blast import NCBIXML
</pre>
</li>
<li>Make a BLAST request using <code>NCBIWWW.qblast("blastn", "nt", seq)</code> method and store results in a file with filename described above.  <code>seq</code> is a variable storing your string sequence. Make sure to close the BLAST request and the save file. Re-open the file for reading. (2 marks)</li>
<li>At this point, you will have an open file containing the BLAST output. Use  <code>NCBIXML.parse(blast_infile)</code> to get the BLAST record. (1 mark)</li>
<li>Based on the title of the first BLAST record alignment (<code>blast_record.alignments[0].title</code>) determine if DNA sequence belongs to fruit fly or grape. Use a list where there is a 1 if the DNA sequence is for the grape and a 0 if it is for the fruit fly.  (1 marks)</li>
<li>Use another list to store the sizes of the alignment matched (<code>blast_record.alignments[0].length</code>). (1 mark)</li>
<li>Make sure to close the input file storing the BLAST result. (0.5 marks)</li>
<li>After all sequences in the DNA <code>input.txt</code> file are processed, print out the list containing 1 if the DNA sequence is for the grape and a 0 if it is for the fruit fly.  Also print out the list storing the sizes of the alignments. (1 mark)</li>
<li>Make sure to close the <code>input.txt</code> file when done. (0.5 marks)</li>
<li>Create a histogram with 50 bins that displays the alignment sequence sizes (from sizes list). Use <code>import matplotlib.mlab</code>. (1 mark)</li>
<li>Create a bar chart that shows how many of the sequences were for the fruit fly and for the grape. Use <code>import matplotlib.pyplot</code>. (2 marks)</li>
<li><strong>Bonus:</strong> Create a different chart (either a different type besides bar/histogram) or use a different Python chart library. (1 mark)</li>
<li><strong>Bonus:</strong> Apply a linear regression or k-means clustering in a useful way to the BLAST query and/or result data. (1 mark)</li>
</ul>
</p>

<h4>Sample Output</h4>

<pre>
Sequence: GGCTGCGGAGACGTTGAAGG Length: 560
Using saved file.
Sequence: CGACAATGCACGACAGAGGA Length: 1190
Using saved file.
Sequence: TTCTCTGTCTTATATTACCG Length: 1470
Using saved file.
Sequence: ACTATTCCAAACAGCTCATA Length: 910
Using saved file.
Sequence: GGACACCGCATTAGCCCAAA Length: 490
Using saved file.
Sequence: GGGTATAATTTCAAAGGAGG Length: 1960
Using saved file.
Sequence: TTATTATATTATTATATTAT Length: 1610
Using saved file.
Sequence: ACAGTGATGCGTGGAGTGCG Length: 307
Using saved file.
Sequence: TAATTTTCTCACGCGCAAAT Length: 1610
Using saved file.
Sequence: CCAAATCAGATTGTGTAACG Length: 1610
Using saved file.
[1, 0, 0, 0, 1, 0, 0, 1, 0, 0]
[1938, 23011485, 24543538, 23542271, 954, 33246, 1348131, 307, 159272, 3436]
</pre>
<img src="img/lab6q1_output.png">

<!--
<h3>Question #2 (5 marks)</h3>

<p>Create a Python program that performs a linear regression and k-means clustering on a data set. Details:

<ul>
<li>Put a comment at the top of the Python file called <code>lab6q2</code> with your name and student number. (0.5 mark)</li>
<li>Download the data file called <a href="lab6q2_input.txt">lab6q2_input.txt</a> and place it in your local directory for reading.</li>
<li>Read each line of the data file into a list. (0.5 mark)</li>
<li>Perform a linear regression using <code>scipy</code>. (2 marks)</li>
<li>Perform a k-means clustering using <code>scipy</code>. (2 marks)</li>
</ul>

<h4>Sample Output</h4>

<pre>
</pre>
-->

<!-- <p>When complete, submit your two code files using Connect: -->

<p>When complete, submit your code file using Connect:
<ul>
<li><code>lab6q1</code> - Python code for question 1.  May be a <code>.txt</code>, <code>.py</code>, or <code>.ipynb</code> file.</li>
<!-- <li><code>lab6q2</code> - Python code for question 2.  May be a <code>.txt</code>, <code>.py</code>, or <code>.ipynb</code> file.</li> -->
</ul>
</p>

<hr>
<img src="../../../../images/go-upwd.gif" alt="*"  align="top" height="14" width="14">
        <a href="../index.html">Home</a>
<hr>

</body>
</html>
